### DISCUSSION TITLE
Middleware in Bot framework composer 2.x

### DISCUSSION THREAD

**Discussion by ic202** [microsoft/BotFramework-Composer#7894] [Q&A]
Is there an example for programming Middleware with Version 2.x?
I was picking something fruit but it doesn't work:

             services.AddSingleton(sp => new QnAMiddleware(sp.GetService<UserState>()));
            services.AddSingleton<IBotFrameworkHttpAdapter, AdapterWithErrorHandler>();

public class AdapterWithErrorHandler : CloudAdapter
    {
        public AdapterWithErrorHandler(IConfiguration configuration, IHttpClientFactory httpClientFactory, ILogger<IBotFrameworkHttpAdapter> logger, QnAMiddleware qnAMiddleware, ConversationState conversationState = default)
            : base(configuration, httpClientFactory, logger)
        {
            if (qnAMiddleware == null)
            {
                throw new NullReferenceException(nameof(qnAMiddleware));
            }
            
            Use(qnAMiddleware);

        }
    }

 public class QnAMiddleware : IMiddleware
    {
        public QnAMiddleware(UserState userState)
        {
           
        }
        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
        {
            await next(cancellationToken);
        }
    }

thank you

**Comment by AntonioChiancone** [comment]
Hi, in Startup.cs > public void ConfigureServices(IServiceCollection services)

            services.AddSingleton<IMiddleware>(_ => new QnAMiddleware());
